### DISCUSSION TITLE
Using rel32 addressing when calling cfuncs, instead of r64

### DISCUSSION THREAD

**Discussion by 64kramsystem** [tenderlove/tenderjit#92] [Ideas]
The current `Runtime#call_cfunc` uses 64-bit addressing:

```rb
@fisk.mov(@fisk.rax, @fisk.uimm(func_loc))
  .call(@fisk.rax)
```

This clobbers RAX, and makes it impossible to perform some calls, e.g. libc's `printf`, which uses RAX to count the floating point parameters (ie. `xmm` registers used).

I'm not familiar with the subject, but this seems to be required by the assumption of a large/huge [memory model](https://eli.thegreenplace.net/2012/01/03/understanding-the-x64-code-models).

In `ISEQCompiler` though, a 32-bits relative address is used for jumps (see operations using `@fisk.absolute`), which assumes a small memory model.

Are there cases where 64-bit addressing is used for cfuncs? If there isn't, `call_cfunc` could be simplified. I've actually tested on a branch if the change works, and it does, however, I don't know if there are other situations where this wouldn't work.

**Comment by tenderlove** [comment]
AFAIK there aren't any cases where we actually need 64 bit addressing, though I haven't confirmed.  I think if the tests pass with your PR we should just merge it?  I don't really see any downsides 😄

**Reply by 64kramsystem** [reply]
Hi there!

My reasoning is that the allocated buffers may reside anywhere in the address space, so that, if a buffer is located too far (GBs) from the function behind called, 32-bit relative addressing would fail.

However, I have no experience in operating systems, so this is 100% speculation :grimacing: 

I'll follow up with my latest work packaged in a PR, including this change.
